### PR TITLE
Improve performance in Password reset flow

### PR DESCRIPTION
**Pull request by ashensw**
### Proposed changes in this pull request

- Improve performance in Password reset flow
- Resolves wso2/product-is#2636

### When should this PR be merged

- Immediately

### Follow up actions

- N/A

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java [hunk 1]
```diff
@@ -259,23 +260,32 @@ private boolean revokeTokens(String username, UserStoreManager userStoreManager)
...
+
+            for (String scope : scopes) {
+
                 AccessTokenDO scopedToken = null;
                 try {
                     // retrieve latest access token for particular client, user and scope combination if its ACTIVE or EXPIRED
                     scopedToken = OAuthTokenPersistenceFactory.getInstance().getAccessTokenDAO()
                             .getLatestAccessToken(clientId, authenticatedUser, userStoreDomain,
-                                    OAuth2Util.buildScopeString(accessTokenDO.getScope()), true);
+                                    scope, true);
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: Can move to previous line

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java [hunk 2]
```diff
@@ -259,23 +260,32 @@ private boolean revokeTokens(String username, UserStoreManager userStoreManager)
...
                 } catch (IdentityOAuth2Exception e) {
                     String errorMsg = "Error occurred while retrieving latest " +
                             "access token issued for Client ID : " +
                             clientId + ", User ID : " + authenticatedUser + " and Scope : " +
-                            OAuth2Util.buildScopeString(accessTokenDO.getScope());
+                            scope;
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: Can move to previous line

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java [hunk 3]
```diff
@@ -259,23 +260,32 @@ private boolean revokeTokens(String username, UserStoreManager userStoreManager)
...
                         OAuth2Util.buildScopeString(accessTokenDO.getScope()));
                 OAuthUtil.clearOAuthCache(accessTokenDO.getConsumerKey(), accessTokenDO.getAuthzUser());
                 OAuthUtil.clearOAuthCache(accessTokenDO.getAccessToken());
+
+                // Get unique scopes list
+                scopes.add(OAuth2Util.buildScopeString(accessTokenDO.getScope()));
+            }
+
+            for (String scope : scopes) {
+
```
Comment (denuwanthi) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: shall we remove unnecessary new lines in several places?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java [hunk 4]
```diff
@@ -259,23 +260,32 @@ private boolean revokeTokens(String username, UserStoreManager userStoreManager)
...
             for (AccessTokenDO accessTokenDO : accessTokenDOs) {
-                //Clear cache
+                // Clear cache
                 OAuthUtil.clearOAuthCache(accessTokenDO.getConsumerKey(), accessTokenDO.getAuthzUser(),
                         OAuth2Util.buildScopeString(accessTokenDO.getScope()));
                 OAuthUtil.clearOAuthCache(accessTokenDO.getConsumerKey(), accessTokenDO.getAuthzUser());
                 OAuthUtil.clearOAuthCache(accessTokenDO.getAccessToken());
+
+                // Get unique scopes list
+                scopes.add(OAuth2Util.buildScopeString(accessTokenDO.getScope()));
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: If there are same set of scopes for multiple access tokens, HashSet won't allow multiple entries with same value. So seems like, there can be situation that all the ACTIVE and EXPIRED tokens will not get REVOKED from database (If there are multiple access tokens with same scopes, only the latest access token will moved to REVOKED state). WDYT?

Comment (GayanM) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: I guess given user, given scope, given client id only one access token exist

Comment (IndunilRathnayake) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: For given user, given scope, given client id, only one ACTIVE token exists, there can be EXPIRED tokens. Previously we move all the ACTIVE and EXPIRED access tokens to REVOKED state [1]. So with this fix, it'll be changed to, move only the latest (ACTIVE or EXPIRED) access token into REVOKED state, for tokens with same scopes.

[1] https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java#L254

Comment (GayanM) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: good catch! but I don't see a major impact. Let's have a discussion

Comment (ashensw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: @IndunilRathnayake This has not changed the previous behaviour IMO. As for the previous implementation also we only revoke the latest access token for a given **user, client id, scope** combination as we are using the getLatestAccessToken method as [here](https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/822/files#diff-0daa59162506998f3129964e2307f192R282).

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java (hunk 3 above)
Comment (ashensw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: Fixed and updated the PR.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java (hunk 1 above)
Comment (ashensw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: Fixed and updated the PR.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java (hunk 2 above)
Comment (ashensw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: Fixed and updated the PR.

**Pull request merged**